text/0026-sassafras-consensus.md

```diff
@@ -0,0 +1,988 @@
+
+Validators generate and submit their candidate tickets along with validity
+proofs to the blockchain. The validity proof ensures that the tickets are
+legitimate while maintaining the anonymity of the authorship.
+
+### 4.2. Validation of Candidate Tickets
+
+Submitted candidate tickets undergo a validation process to verify their
+authenticity, integrity and compliance with other protocol-specific rules.
+
+### 4.3. Tickets and Slots Binding
+
+After collecting all candidate tickets, a deterministic method is employed to
+associate some of these tickets with specific slots.
+
+### 4.4. Claim of Ticket Ownership
+
+Validators assert ownership of tickets during the block production phase. This
+step establishes a secure binding between validators and their respective slots.
+
+### 4.5. Validation of Ticket Ownership
+
+During block verification, the claims of ticket ownership are rigorously
+validated to uphold the protocol's integrity.
+
+
+## 5. Bandernatch VRFs Cryptographic Primitives
+
+This chapter provides a high-level overview of the Bandersnatch VRF primitive as
+it relates to the Sassafras protocol.
+
+It's important to note that this section is not intended to serve as an
+exhaustive exploration of the mathematically intensive foundations of the
+cryptographic primitive. Instead, its primary purpose is to offer a concise and
+comprehensible interpretation of the primitive within the context of this RFC.
+
+For a more detailed and mathematical understanding of Ring-VRFs, we recommend
+referring to the Ring-VRF research [paper](https://eprint.iacr.org/2023/002.pdf).
+
+### 5.1. VRF Input
+
+The VRF Input, denoted as `VrfInput`, is constructed using a domain and some
+arbitrary data using the `vrf_input(domain: OCTET_STRING, buf: OCTET_STRING)`
+function. This function is left opaque here and can be read in the actual
+reference implementation.
+
+The VRF Input, denoted as `VrfInput`, is constructed by combining a domain identifier
+with arbitrary data using the `vrf_input` function:
+
+```rust
+    fn vrf_input(domain: OCTET_STRING, buf: OCTET_STRING) -> VrfInput;
+```
+
+While the specific implementation details of this function are intentionally
+omitted here, you can find the complete implementation in the
+[`bandersnatch_vrfs`](https://github.com/w3f/ring-vrf/blob/18614458ca4cb335c88d4e710c13906a76f51e43/bandersnatch_vrfs/src/lib.rs#L57) 
+reference implementation.
+
+Helper function to construct a `VrfInput` from a sequence of `data` items:
+
+```rust
+    fn vrf_input_from_items(domain: OCTET_STRING, data: SEQUENCE_OF OCTET_STRING) -> VrfInput {
+        buf = OCTET_STRING(SIZE(0));
+        for item in data {
+            buf.append(item);
+            buf.append(length(item) as U8);
+        }
+        return vrf_input(domain, buf);
+    }
+```
+
+### 5.2. VRF Output
+
+Given a `VrfInput` object, the corresponding `VrfOutput`, also referred to as
+`VrfPreOutput`, is computed using a Bandersnatch secret key.
```

Comment:
I only call this `VrfPreOutput` since I wanted to deprecate the `VrfOutput` being used here.  Also `vrf_output` is now a different function.

Alone `preoutput` is pretty damn random, but it's algebraic so it carries "interesting" relationships between different uses, especially related keys.